src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -210,16 +224,55 @@ protected override async Task OnRun(CancellationToken token)
                         // Complete the pipeline since the action list is only executed once
                         // for collection rules with startup triggers.
                         completePipeline = true;
+
+                        _stateHolder.StartupTriggerCompleted();
                     }
                 }
             }
             catch (OperationCanceledException) when (durationCancellationSource.IsCancellationRequested)
             {
                 // This exception is caused by the pipeline duration expiring.
                 // Handle it to allow pipeline to be in completed state.
+                _stateHolder.RuleDurationReached();
+            }
+            catch (Exception ex)
+            {
+                _stateHolder.RuleFailure(ex.Message);
+                throw;
             }
         }
 
+        public Queue<DateTime> DequeueOldTimestamps(TimeSpan? actionCountWindowDuration, DateTime currentTimestamp)
```

Comment:
I know there was some discussion earlier about not wanting to have a copy of `_executionTimestamps` in `CollectionRuleService`; is there any issue with having this method interact directly with `_executionTimestamps` (instead of returning a copy), and then `CheckForThrottling` directly using its size? I think there had been concerns about updating the `CollectionRuleStateHolder` from the front-end, but is there any reason (from the perspective of needing to lock resources) that we couldn't do it this way for the timestamps?

Reply:
Move to `CollectionRuleStateHolder` (holds a copy)